### PR TITLE
1/n Gemini Update deserialize to take data in the form of completion params

### PR DESCRIPTION
**Pull request by Ankush-lastmile**
1/n Gemini Update deserialize to take data in the form of completion params

## What

Updated deserialize() in Gemini Model Parser to take in completion data from prompt input if it is already in the form of the Gemini Completion data

Now, you can specify input in the form of a request to gemini api, ie the parts.

## Why

Previously This was skipped due to a lack of understanding of how the api requests worked.

## Testplan
Validating 2 pieces of functionality with this test plan:
1. setting the generation config for Gemini in setings. Generation Config is gemini's way of defining settings like topP
2. Deserialize() can take in data just like the api, by setting "contents" in the prompt input.
testplan

testplan 2

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/lastmile-ai/aiconfig/pull/532).
* #558
* #534
* #531
* __->__ #532

### CODE REVIEW

File: extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py [hunk 1]
```diff
@@ -134,17 +135,34 @@ async def deserialize(
...
+
 
         # Build Completion data
         model_settings = self.get_model_settings(prompt, aiconfig)
 
         completion_data = refine_chat_completion_params(model_settings)
 
+        
+
         messages = self._construct_chat_history(prompt, aiconfig, params)
```
Comment (rossdanlm) [inline extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:147]: `construct_prior_chat_history`

File: extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py [hunk 2]
```diff
@@ -134,17 +135,34 @@ async def deserialize(
...
+
+        if contains_prompt_template(prompt):
+            resolved_prompt = resolve_prompt(prompt, params, aiconfig)
+            messages.append({"role": "user", "parts": [{"text": resolved_prompt}]})
+            completion_data["contents"] = messages
+        else:
+            prompt_input = prompt.input
+            # This is checking attributes and not a dict like object. in schema.py, PromptInput allows arbitrary attributes/data, and gets serialized as an attribute because it is a pydantic type
+            if not hasattr(prompt_input, "contents"):
+                raise ValueError("Unable to deserialize input. Prompt input type is not a string, Gemini Model Parser expects prompt input to contain a 'contents' field as expected by Gemini API")
```
Comment (rossdanlm) [inline extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:158]: Link to API doc url

File: extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py [hunk 3]
```diff
@@ -292,4 +310,47 @@ def refine_chat_completion_params(model_settings):
     return completion_data
 
 
+def parameterize_supported_gemini_input_data(part: Any, prompt: Prompt, aiconfig: "AIConfigRuntime", input_params: dict[str, Any]):
+    """
+    Parameterizes the input for the Gemini API based on the type of the input part.
+    This function specifically handles string-based types in the context of Gemini API.
+
+    Gemini API supports a variety of inputs, some of which are called 'parts', coined by Gemini API.
```
Comment (rossdanlm) [inline extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:318]: +1

File: extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py [hunk 4]
```diff
@@ -292,4 +310,47 @@ def refine_chat_completion_params(model_settings):
...
+    * Important: This method supports a subset of the Gemini API inputs. Specifically, types that contain strings (and not protos, images, etc)
```
Comment (rossdanlm) [inline extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:319]: pls rename to proto-buff

File: extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py [hunk 5]
```diff
@@ -134,17 +135,34 @@ async def deserialize(
                 "on_deserialize_start", __name__, {"prompt": prompt, "params": params}
             )
         )
-        resolved_prompt = resolve_prompt(prompt, params, aiconfig)
+
```
Comment (rossdanlm) [inline extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:138]: delete extra line?

**Review by rossdanlm — APPROVED** [review]
pls add test plan and address comments
image

File: extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py [hunk 6]
```diff
@@ -134,17 +135,36 @@ async def deserialize(
...
+            messages = self._construct_chat_history(prompt, aiconfig, params)
 
-        completion_data["contents"] = messages
+            resolved_prompt = resolve_prompt(prompt, params, aiconfig)
+            
+            messages.append({"role": "user", "parts": [{"text": resolved_prompt}]})
+            
+            completion_data["contents"] = messages
+        else:
+            # If contents is already set, do not construct chat history. TODO: @Ankush-lastmile
```
Comment (Ankush-lastmile) [inline extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:153]: If contents is already set (ie: defined as a key in a dict --> see 3rd and 4th case in "supported types" below)

File: extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py [hunk 7]
```diff
@@ -134,17 +135,36 @@ async def deserialize(
...
-        resolved_prompt = resolve_prompt(prompt, params, aiconfig)
 
         # Build Completion data
         model_settings = self.get_model_settings(prompt, aiconfig)
 
         completion_data = refine_chat_completion_params(model_settings)
 
-        messages = self._construct_chat_history(prompt, aiconfig, params)
-        messages.append({"role": "user", "parts": [{"text": resolved_prompt}]})
+        if contains_prompt_template(prompt):
```
Comment (Ankush-lastmile) [inline extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py:144]: add comment & github issue to construct chat history with a contents dict input

**Comment by Ankush-lastmile** [comment]
discussed and resolved other comments offline cc: @rossdanlm

**Pull request merged**